mlflow/tensorflow/__init__.py

```diff
@@ -769,6 +787,39 @@ def train(original, self, *args, **kwargs):
 
         result = original(self, *args, **kwargs)
 
+        def _extract_data_from_input_fn():
+            def _extract_dict_or_tensor(next_input):
+                input_ex_dict_tensor = None
+                if isinstance(next_input, dict):
+                    input_ex_dict_tensor = {
+                        k: _extract_tensor_as_numpy_input_example_slice(v)
+                        for k, v in next_input.items()
+                    }
+                elif isinstance(next_input, tensorflow.Tensor):
+                    input_ex_dict_tensor = _extract_tensor_as_numpy_input_example_slice(next_input)
+                return input_ex_dict_tensor
+
+            input_training_data = kwargs.get("input_fn")()
```

Comment:
Just making sure - is it always safe to call this method multiple times?

Reply:
the output of `input_fn' can be of type [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) or [A tuple (features, labels)](https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator#train). So it should be safe as there should be no side-effects for getting the return of those repeatedly.

Reply:
That being said, if those datasets are generated with some custom code with transient side-effects inside `input_fn` , that could be an issue. Just to be safe, we could do call that function once and wrap the results inside another function (with no side-effects) that can be passed down to the train, but we will have the same [issue](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/autologging_utils/safety.py#L869) that GeneratorType change was throwing: AssertionError around the original input_fn being mutated.

Reply:
This seems okay to me for now :)